File: backend/python/heartbeat.py
import json
import time
from datetime import datetime
from pytz import timezone

from db import db_timezones, db_notifications

TIME_FORMAT = "%H:%M:%S"


async def heartbeat():
    return await tick_update(db_timezones(), db_notifications())


async def tick_update(timezones, notifications):
    ts = time.time()

    for notification in notifications:
        update_notification_status(ts, notification)
        update_notification_time(notification)

    for timezone_ in timezones:
        update_timezones_time(ts, timezone_)

    response = {
        'utc_time':
        datetime.fromtimestamp(ts, timezone('UTC')).strftime(TIME_FORMAT),
        'timezones':
        timezones,
        'notifications':
        notifications
    }

    return json.dumps(response)


def update_notification_time(notification):
    time_zone = notification['time_zone']
    timestamp = notification['timestamp']
    notification['time'] = datetime.fromtimestamp(
        timestamp, timezone(time_zone)).strftime(TIME_FORMAT)


def update_notification_status(time, notification):
    if (notification['timestamp'] < time):
        notification['status'] = 'FINISHED'
    elif ((notification['timestamp'] - time) > 5):
        notification['status'] = 'PENDING'
    elif ((notification['timestamp'] - time) < 5
          and (notification['timestamp'] - time) > 1):
        notification['status'] = 'GET_READY'
    elif ((notification['timestamp'] - time) < 1):
        notification['status'] = 'ACTIVE'


def update_timezones_time(time, time_zone):
    timezone_ = time_zone['time_zone']
    time_zone['timestamp'] = time
    time_zone['time'] = datetime.fromtimestamp(
        time, timezone(timezone_)).strftime(TIME_FORMAT)

File: backend/node_js/heartbeat.js
const path = require('path');
const db = require( path.resolve( __dirname, "./db.js" ) );




//update notifications and timezones status' every second
const tick_update = () => {
	const timezones = db.timezones()
	const notifications = db.notifications()
    const ts = Date.now() / 1_000
    notifications.map((notification) => {
        update_notification_status(ts, notification)
        update_notification_time(notification)
    })
    timezones.map((timezone_) => {

        update_timezones_time(ts, timezone_) 
    })
    const response = {'utc_time': new Date(Date.now()).toLocaleTimeString('gb-GB', {timeZone: 'UTC'}),
                'timezones': timezones,
                'notifications': notifications}
    return JSON.stringify(response)
}

const update_notification_time = (notification) =>{
    const time_zone_str = notification['time_zone']
    const timestamp = notification['timestamp']
    //https://en.wikipedia.org/wiki/List_of_tz_database_time_zones
    notification['time'] = new Date(timestamp * 1_000).toLocaleTimeString('gb-GB', {timeZone: time_zone_str})
}

const update_notification_status =(time, notification) => {
    if(notification['timestamp'] < time)
        notification['status'] = 'FINISHED'
    else if((notification['timestamp'] - time) > 5)
        notification['status'] = 'PENDING'
    else if((notification['timestamp'] - time) < 5 && (notification['timestamp'] - time) > 1)
        notification['status'] = 'GET_READY'
    else if((notification['timestamp'] - time) < 1 )
        notification['status'] = 'ACTIVE'
}

const update_timezones_time = (time_actual, time_zone) => {
    const timezone_str = time_zone['time_zone']
    time_zone['timestamp'] = time_actual
    time_zone['time'] = new Date(time_actual* 1_000).toLocaleTimeString('gb-GB', {timeZone: timezone_str})
} 

module.exports = {
    heartbeat: function () {
        return tick_update()
    }
  };
